### ISSUE TITLE
Remove a PowerLaw background from EELS

### ISSUE THREAD

**Issue by WolfgangKoe** [hyperspy/hyperspy#1844]
Hi there,

I have an EEL-dataset with a sulfur and carbon edge. I'm trying to remove the PowerLaw background in order to get the "pure" signal from the sulfur and the carbon. I was looking for some ideas, but I found nothing working in the documentation or the issues. If there is some easy solution for this, please tell me. I'm sure there is one since I'm not the first person facing this problem ;-).

My first attempts went like this:

```
import hyperspy.api as hs
import numpy as np
import matplotlib.pyplot as plt

EELS_data = hs.load('file')

EELS_data.set_microscope_parameters(beam_energy=60,
                                              convergence_angle=0.2,
                                              collection_angle=2.55
                                              )
m = EELS_data.isig[135.:160.].create_model()

m.fit()

m.plot(plot_components=True)
m.print_current_values()

```

The fit is fine, since the 1D Basesignal model already contains the PowerLaw. But in this way I created a model of EELS_data.isig[135.:160.] and not of EELS_data, were I want to subtract the calculated signal of the fit from the signal dimension of the data over the full range. 
And at this point I didn't find a way to go the next steps. I first need to expand the fit and afterwards substract it from my original signal. In the last step I want to plot it of course.

Maybe someone can help me.

Thx

**Comment by francisco-dlp** [comment]
Usually we don't subtract the background from the raw data when fitting an EELS model because it usually results in a less accurate fit. Still, you can remove it using ``EELS_data.remove_background()``. Or is it from the model that you want to subtract the background? If that's the case, simply deactivating it (after a successful fit) should do:

```python

m.components.PowerLaw.active = False

``` 

Or maybe I didn't understand you question?

**Comment by WolfgangKoe** [comment]
Thank you for your answer. Well, I somehow want to remove the background from the signal I‘m interested in. I tried `EELS-data.remove_background()` in the first place. But it didn‘t work and I couldn’t actually find out how this function works. If it just works after fitting the data, because you fit the background as well, than it is clear ;-). But in this case I need a proper model for the sulfur and carbon, what I don‘t have ;-). 

Basically I‘m looking for some kind of a hyperspy implementation of the 3 window method. Maybe this is the better question...

**Comment by francisco-dlp** [comment]
If ``remove_background`` didn't work is probably because you don't have any of the ``gui`` packages installed. You can try installing [hyperspy_gui_ipywidgets](https://github.com/hyperspy/hyperspy_gui_ipywidgets). Alternatively, you can still use it non-interactively

```python
s.remove_background(signal_range=(100, 200))
```

**Comment by WolfgangKoe** [comment]
I will give it a try next week and will report. Thx so far...

**Comment by thomasaarholt** [comment]
We should clarify in the docs whether it matters if the signal_range is specified as ints and floats.

**Comment by WolfgangKoe** [comment]
Hi there,

I installed the `gui` packages with anaconda. Here is the protocol just that you know what happened.

```
conda install hyperspy-gui-ipywidgets -c conda-forge
Fetching package metadata ...........
Solving package specifications: .

Package plan for installation in environment /home/wolfgangkoentges/Programme/anaconda3:

The following packages will be UPDATED:

    conda:     4.3.22-py36_0 --> 4.3.34-py36_0 conda-forge

The following packages will be SUPERSEDED by a higher-priority channel:

    conda-env: 2.6.0-0       --> 2.6.0-0       conda-forge

Proceed ([y]/n)? y

conda-env-2.6. 100% |################################| Time: 0:00:00 575.27 kB/s
conda-4.3.34-p 100% |################################| Time: 0:00:01 483.49 kB/s
```

This was really the problem, why `remove_background()` didn't do anything. Now at least something happens ;-). But if I run the script through my data I get the following error, after I chose to remove a PowerLaw background with `EELS_data.remove_background()`:

```
Traceback (most recent call last):
  File "/home/wolfgangkoentges/Programme/anaconda3/lib/python3.6/site-packages/traitsui/qt4/ui_base.py", line 405, in done
    elif self._ok_to_close(bool(r)):
  File "/home/wolfgangkoentges/Programme/anaconda3/lib/python3.6/site-packages/traitsui/qt4/ui_base.py", line 422, in _ok_to_close
    ok_to_close = self._ui.handler.close(self._ui.info, is_ok)
  File "/home/wolfgangkoentges/Programme/anaconda3/lib/python3.6/site-packages/hyperspy_gui_traitsui/tools.py", line 26, in close
    self.apply(info)
  File "/home/wolfgangkoentges/Programme/anaconda3/lib/python3.6/site-packages/hyperspy_gui_traitsui/tools.py", line 41, in apply
    obj.apply()
  File "/home/wolfgangkoentges/Programme/anaconda3/lib/python3.6/site-packages/hyperspy/signal_tools.py", line 764, in apply
    polynomial_order=self.polynomial_order)
  File "/home/wolfgangkoentges/Programme/anaconda3/lib/python3.6/site-packages/hyperspy/_signals/signal1d.py", line 1044, in remove_background
    " not recognized")
ValueError: Background type: Power Law not recognized

If you suspect this is an IPython bug, please report it at:
    https://github.com/ipython/ipython/issues
or send an email to the mailing list at ipython-dev@python.org

You can print a more detailed traceback right now with "%tb", or use "%debug"
to interactively debug it.

Extra-detailed tracebacks for bug-reporting purposes can be enabled via:
    %config Application.verbose_crash=True
```

It didn't print more information. I cannot say, if this is a IPython bug, since it just says, that no Power Law background is recognized. But there is obviously one in the data. When I reduce the window to 

`EELS_data_dic.isig[135.:160.]`

where you definitely see only the Power Law background, then I get the same error. Maybe I missed a package or there is still an error in my script?

**Comment by thomasaarholt** [comment]
You should paste in the code you ran :)

My guess is that you gave the argument `background_type='Power Law'`, instead of 'PowerLaw' (note the lack of a space).

**Comment by WolfgangKoe** [comment]
The code is basically the same as before:

```
import hyperspy.api as hs
import numpy as np
import matplotlib.pyplot as plt

EELS_data = hs.load('file')

EELS_data.remove_background()
```
But it is possible, that you are right. In the interactive menu, that pops up, I just can choose 'Power Law'. I saw this comment from you in another issue ;-).

**Comment by thomasaarholt** [comment]
Huh, I'm partially right. I just tried your code and it worked fine:
![image](https://user-images.githubusercontent.com/2721423/36661677-441996c6-1adc-11e8-8ca9-849b312fcbb3.png)

However, if I specify the background type and the signal range, it does not work with 'Power Law' but does work with 'PowerLaw'. My only guess is that it has something to do with the qt4 code that you are quoting in your error.

### Quick fix
Would you mind trying specifying the limits and type? Something like this (and note that it returns a NEW signal when you do it like this):
```python
EELS_data = hs.load('file')
EELS_data_background_subtracted = EELS_data.remove_background(signal_range=(10.,15.), background_type='PowerLaw')
EELS_data_background_subtracted.plot()
```
### Fix the problem properly (maybe)
In addition, could you try turning the traitsui GUI off? (you don't need it if you have installed the ipywidgets one.)
Do the following:
```
hs.preferences.gui()
```
Now select the tab "GUIs", and then tick so that the "Enable traitsui gui" is off.

From now on, to display plots, restart the kernel in the jupyter notebook (Kernel -> Restart), and lead all new notebook files with the following line:

`%matplotlib notebook`

So the example from the beginning becomes:

```python
%matplotlib notebook
import hyperspy.api as hs
EELS_data = hs.load('file')

EELS_data.remove_background()
```

I *hope* that this fixes it.

**Comment by WolfgangKoe** [comment]
I'm not using the jupyter notebook, but starting the script "Remove_Background_Test.py" in the terminal. I just tried a dataset from the eelsdb from the documentation. 
If I use:

```
import hyperspy.api as hs

EELS_data = hs.load('file')

EELS_data.remove_background()
```

Then I choose 'Power Law' and I still get:

```
$ ipython Remove_Background_Test.py
Using matplotlib backend: Qt5Agg
Traceback (most recent call last):
  File "/home/wolfgangkoentges/Programme/anaconda3/lib/python3.6/site-packages/traitsui/qt4/ui_base.py", line 405, in done
    elif self._ok_to_close(bool(r)):
  File "/home/wolfgangkoentges/Programme/anaconda3/lib/python3.6/site-packages/traitsui/qt4/ui_base.py", line 422, in _ok_to_close
    ok_to_close = self._ui.handler.close(self._ui.info, is_ok)
  File "/home/wolfgangkoentges/Programme/anaconda3/lib/python3.6/site-packages/hyperspy_gui_traitsui/tools.py", line 26, in close
    self.apply(info)
  File "/home/wolfgangkoentges/Programme/anaconda3/lib/python3.6/site-packages/hyperspy_gui_traitsui/tools.py", line 41, in apply
    obj.apply()
  File "/home/wolfgangkoentges/Programme/anaconda3/lib/python3.6/site-packages/hyperspy/signal_tools.py", line 764, in apply
    polynomial_order=self.polynomial_order)
  File "/home/wolfgangkoentges/Programme/anaconda3/lib/python3.6/site-packages/hyperspy/_signals/signal1d.py", line 1044, in remove_background
    " not recognized")
ValueError: Background type: Power Law not recognized

If you suspect this is an IPython bug, please report it at:
    https://github.com/ipython/ipython/issues
or send an email to the mailing list at ipython-dev@python.org

You can print a more detailed traceback right now with "%tb", or use "%debug"
to interactively debug it.

Extra-detailed tracebacks for bug-reporting purposes can be enabled via:
    %config Application.verbose_crash=True

```

When I apply the Quick fix, I get what I wanted!

```
EELS_data = hs.load('file')
EELS_data_background_subtracted = EELS_data.remove_background(signal_range=(130.,160.), background_type='PowerLaw')
EELS_data_background_subtracted.plot()
```

Without removing background:

[Sulfur_Carbon_edge_background.pdf](https://github.com/hyperspy/hyperspy/files/1758401/Sulfur_Carbon_edge_background.pdf)

With removing background using the quick fix:

[Sulfur_Carbon_edge_w-o_background.pdf](https://github.com/hyperspy/hyperspy/files/1758407/Sulfur_Carbon_edge_w-o_background.pdf)

**Comment by francisco-dlp** [comment]
Just like @thomasaarholt, I can't reproduce the issue with the following script:

```python
import hyperspy.api as hs
s, = hs.datasets.eelsdb(title="Boron Nitride", author="Bertoni", spectrum_type="coreloss")
s.remove_background()
```

Which version of HyperSpy are you using?

**Comment by thomasaarholt** [comment]
~~@francisco-dlp The main problem is that @WolfgangKoe is calling hyperspy in a non-interactive manner, while the s.remove_background() function anticipates the GUI in order to remove the background.~~

Sorry, just realised that @WolfgangKoe is calling ipython, not python.

**Comment by WolfgangKoe** [comment]
Does this mean, that I can't use the interactive functions of hyperspy without using jupyter notebook?

**Comment by francisco-dlp** [comment]
No, it should work. The following script works for me, both using ipython and plain python to run it:

```python
import matplotlib.pyplot as plt
import hyperspy.api as hs

s, = hs.datasets.eelsdb(title="Boron Nitride", author="Bertoni", spectrum_type="coreloss")
s.remove_background()
plt.show()
```

**Comment by WolfgangKoe** [comment]
I tried your code and it should work, but it yields to the error I mentioned earlier...

**Comment by francisco-dlp** [comment]
Which is your HyperSpy version?

**Comment by WolfgangKoe** [comment]
```
hyperspy                  1.3                      py36_0    conda-forge
hyperspy-gui-ipywidgets   1.0.0                    py36_0    conda-forge
hyperspy-gui-traitsui     1.0.3                     <pip>

```

**Comment by thomasaarholt** [comment]
Could this be a qt5 issue (@WolfgangKoe indicates he's using qt5)?

Running your code I'm getting a TkAgg warning here on windows. 
```
WARNING:hyperspy_gui_traitsui:The TkAgg matplotlib backend is not supported by the installed traitsui version and the ETS toolkit has been set to null. To set the ETS toolkit independently from the matplotlib backend,
```

**Comment by francisco-dlp** [comment]
@WolfgangKoe, could you try printing your hyperspy version from inside the script?

```python

import matplotlib.pyplot as plt
import hyperspy.api as hs
import hyperspy

print(hyperspy.__version__)

s, = hs.datasets.eelsdb(title="Boron Nitride", author="Bertoni", spectrum_type="coreloss")
s.remove_background()
plt.show()

**Comment by WolfgangKoe** [comment]
```
$ ipython Remove_Background_Test.py 
Using matplotlib backend: Qt5Agg
1.3
```
And then comes the error again...

**Comment by francisco-dlp** [comment]
I think that you're being hit by #1625 but the puzzling thing is that this bug was fixed in v1.3; hence this bug shouldn't affect you. My bet is that somehow your installation is pointing to the old version of [signal_tools.py](https://github.com/hyperspy/hyperspy/blob/RELEASE_next_minor/hyperspy/signal_tools.py#L760).

**Comment by WolfgangKoe** [comment]
Okay, how can I check this? This will probably a task of tomorrow, since I have to leave my office now ;-).

**Comment by thomasaarholt** [comment]
It is indeed very odd.

You may find that an easy fix is uninstalling and reinstalling hyperspy.

**Comment by thomasaarholt** [comment]
Wolfgang has Qt5 installed. Last I checked (yesterday), this doesn't work with hyperspy (even with the current latest pyface) on Windows. Does it work on Linux/Mac (whatever you're using)?

**Comment by francisco-dlp** [comment]
Yes, I am on Linux. But this issue should not depend on the installed qt nor on the platform: it seems to be an old bug that has somehow resurfaced for some. @thomasaarholt, if it doesn't work for you either, could you have  a look at the [signal_tools line concerned in your installation](https://github.com/hyperspy/hyperspy/blob/RELEASE_next_minor/hyperspy/signal_tools.py#L760) ?

**Comment by ericpre** [comment]
I have just tested it on windows (winpython with pyqt5) on release 1.3 and on the next_minor branch, it works both as expected.

**Comment by thomasaarholt** [comment]
First off, I never had @WolfgangKoe's error.

Having reinstalled qt4 and forced the script to use qt4:
```python
# at the top of the script
import matplotlib
matplotlib.use('Qt4Agg',warn=False, force=True)
```
I can now sortof succesfully run it. The plot appears, but I can only see the plot while I'm dragging the red region. Otherwise the plot is blank.

![image](https://user-images.githubusercontent.com/2721423/36725212-db9ec67c-1bb5-11e8-830d-496c9be8741e.png)

**Comment by ericpre** [comment]
This is a different known issue. What version of matplotlib are you using? If I remember correctly, more recent version of matplotlib have issue with qt4 and this is why matplotlib in anaconda depends on qt5.

**Comment by WolfgangKoe** [comment]
So I just tried to reinstall Hyperspy after deinstalling the package. When I'm running my code as before:

```
import matplotlib.pyplot as plt
import hyperspy.api as hs

hs.preferences.gui()

s = hs.datasets.eelsdb(title="Boron Nitride", author="Bertoni", spectrum_type="coreloss")
s[0].remove_background()
plt.show()
```

 I have an error with hyperspy_gui-traitsui:
```

WARNING:hyperspy.api:The traitsui GUI elements are not available, probably because the hyperspy_gui_traitui package is not installed.
---------------------------------------------------------------------------
ValueError                                Traceback (most recent call last)
/media/wolfgangkoentges/D/Remove_Background_Test.py in <module>()
      4 #print(hyperspy.__version__)
      5 
----> 6 paf.hs.preferences.gui()
      7 
      8 s = paf.hs.datasets.eelsdb(title="Boron Nitride", author="Bertoni", spectrum_type="coreloss")

~/Programme/anaconda3/lib/python3.6/site-packages/hyperspy/ui_registry.py in pg(self, display, toolkit, **kwargs)
    172     def pg(self, display=True, toolkit=None, **kwargs):
    173         return get_gui(self, toolkey=toolkey, display=display,
--> 174                        toolkit=toolkit, **kwargs)
    175     return pg
    176 

~/Programme/anaconda3/lib/python3.6/site-packages/hyperspy/ui_registry.py in get_gui(self, toolkey, display, toolkit, **kwargs)
    127                 "manually select a toolkit with the `toolkit` argument.".format(
    128                     _toolkits_to_string(available_disabled_toolkits),
--> 129                     is_or_are, them_or_it)
    130             )
    131 

ValueError: No toolkit available. The ipywidgets toolkit is installed but disabled in `preferences`. Enable them in `preferences` or manually select a toolkit with the `toolkit` argument.

```

How can I enable the ipywidgets manually?

**Comment by francisco-dlp** [comment]
```python
hs.preferences.GUIs.enable_ipywidgets_gui = True
hs.preferences.save()
```

**Comment by thomasaarholt** [comment]
But unless I've seriously misunderstood, the ipywidgets gui will only work in the notebook.

**Comment by WolfgangKoe** [comment]
It seems so. I still get the warning.
`
WARNING:hyperspy.api:The traitsui GUI elements are not available, probably because the hyperspy_gui_traitui package is not installed.
`

what wasn't the case previously. And .remove_background() does not work. The manual way with

`.remove_background(signal_range=(130.,160.), background_type='PowerLaw')`

works as before

**Comment by ericpre** [comment]
Closing this, since the issue with hyperspy installation and the qt backend have been simplified in recent releases. Feel free to reopen if necessary.

**Comment by wcdyang** [comment]
Dear all,

I have trouble performing the `remove_background()`. It was executed as following.

```python
%matplotlib qt
import hyperspy.api as hspy
import matplotlib.pyplot as plt
import numpy as np
```
```python
n2 = hspy.load('n2_offset_high.hdf5')
n2.remove_background()
```
I received a warning as below:

> C:\Users\wdy\anaconda3\envs\hspy\lib\site-packages\hyperspy\drawing\_widgets\range.py:285: MatplotlibDeprecationWarning: 
The buttonDown attribute was deprecated in Matplotlib 3.1 and will be removed in 3.3.
  self.buttonDown = False

It also showed an error:
```
---------------------------------------------------------------------------
AttributeError                            Traceback (most recent call last)
<ipython-input-20-7333eab21dc1> in <module>
      1 n2 = hspy.load('n2_offset_high.hdf5')
----> 2 n2.remove_background()

~\anaconda3\envs\hspy\lib\site-packages\hyperspy\_signals\signal1d.py in remove_background(self, signal_range, background_type, polynomial_order, fast, zero_fill, plot_remainder, show_progressbar, display, toolkit)
   1075                                    plot_remainder=plot_remainder,
   1076                                    show_progressbar=show_progressbar,
-> 1077                                    zero_fill=zero_fill)
   1078             return br.gui(display=display, toolkit=toolkit)
   1079         else:

~\anaconda3\envs\hspy\lib\site-packages\hyperspy\signal_tools.py in __init__(self, signal, background_type, polynomial_order, fast, plot_remainder, zero_fill, show_progressbar)
    672                  fast=True, plot_remainder=True, zero_fill=False,
    673                  show_progressbar=None):
--> 674         super(BackgroundRemoval, self).__init__(signal)
    675         # setting the polynomial order will change the backgroud_type to
    676         # polynomial, so we set it before setting the background type

~\anaconda3\envs\hspy\lib\site-packages\hyperspy\signal_tools.py in __init__(self, signal)
     52         self.span_selector = None
     53         self.signal.plot()
---> 54         self.span_selector_switch(on=True)
     55 
     56     def on_disabling_span_selector(self):

~\anaconda3\envs\hspy\lib\site-packages\hyperspy\signal_tools.py in span_selector_switch(self, on)
     66                     self.signal._plot.signal_plot.ax,
     67                     onselect=self.update_span_selector_traits,
---> 68                     onmove_callback=self.update_span_selector_traits,)
     69 
     70         elif self.span_selector is not None:

~\anaconda3\envs\hspy\lib\site-packages\hyperspy\drawing\_widgets\range.py in __init__(self, ax, **kwargs)
    283         self.step_ax = None
    284         self.bounds_check = False
--> 285         self.buttonDown = False
    286         self.snap_size = False
    287         self.snap_position = False

~\anaconda3\envs\hspy\lib\site-packages\matplotlib\cbook\deprecation.py in __set__(self, instance, value)
    216                         from . import _warn_external
    217                         _warn_external(warning)
--> 218                     return super().__set__(instance, value)
    219 
    220                 def __delete__(self, instance):

AttributeError: can't set attribute
```
Although it gave me an interactive plot, whenever I tried to define the range by clicking inside the axes of the figure and dragging to the right without releasing the button, the background window was not created in the interactive plot and a new warning was showed as below:
```
Traceback (most recent call last):
  File "C:\Users\wdy\anaconda3\envs\hspy\lib\site-packages\matplotlib\cbook\__init__.py", line 216, in process
    func(*args, **kwargs)
  File "C:\Users\wdy\anaconda3\envs\hspy\lib\site-packages\matplotlib\widgets.py", line 1633, in onmove
    self._onmove(event)
  File "C:\Users\wdy\anaconda3\envs\hspy\lib\site-packages\matplotlib\widgets.py", line 1885, in _onmove
    self.onmove_callback(vmin, vmax)
  File "C:\Users\wdy\anaconda3\envs\hspy\lib\site-packages\hyperspy\signal_tools.py", line 78, in update_span_selector_traits
    x0 = self.span_selector.rect.get_x()
AttributeError: 'NoneType' object has no attribute 'rect'
```
The same result was reproduced when the "Enable traitsui gui" was either checked or not.

The packages in the environment can be found here.
[pkgs.txt](https://github.com/hyperspy/hyperspy/files/3287827/pkgs.txt)

Any thoughts? Thank you.

**Comment by jat255** [comment]
@weichang252 this is a known issue (see https://github.com/hyperspy/hyperspy/issues/2195 for details). It has been fixed, but a new version has not been released. The issue is due to a problem with the latest version of `matplotlib`. In the meantime, before a new release of HyperSpy is made, you can downgrade matplotlib to version 3.0 to fix the issue. You should run `conda install -c conda-forge matplotlib=3.0` in your `hspy` environment to workaround the issue.

**Comment by wcdyang** [comment]
@jat255 thank you so much.
